src/digital.rs

```diff
@@ -123,3 +123,112 @@ pub trait InputPin {
     /// Is the input pin low?
     fn try_is_low(&self) -> Result<bool, Self::Error>;
 }
+
+/// Dummy GPIO pin
+///
+/// These structures are useful when using optional pins, for example
+/// when using some SPI devices.
+pub mod dummy {
+    use super::{toggleable, InputPin, OutputPin, StatefulOutputPin};
+    use core::{convert::Infallible, marker::PhantomData};
+
+    /// Dummy input pin level
+    pub mod level {
+        /// Dummy input pin will always behave as being high.
+        pub struct High;
+        /// Dummy input pin will always behave as being low.
+        pub struct Low;
+    }
+
+    /// Dummy (No-op) pin
+    ///
+    /// This will discard any value set to it and when checked always behave
+    /// according to the value provided at construction time
+    pub struct DummyPin<L = level::Low> {
+        _l: PhantomData<L>,
+    }
+
+    impl DummyPin<level::Low> {
+        /// Create new instance
+        ///
+        /// When checked it will always behave as being low
+        pub fn new() -> Self {
+            DummyPin { _l: PhantomData }
+        }
+
+        /// Create new instance
+        ///
+        /// When checked it will always behave as being low
+        pub fn new_low() -> Self {
+            DummyPin { _l: PhantomData }
+        }
+    }
+
+    impl DummyPin<level::High> {
+        /// Create new instance
+        ///
+        /// When checked it will always behave as being high
+        pub fn new_high() -> Self {
+            DummyPin { _l: PhantomData }
+        }
+    }
+
+    impl<L> OutputPin for DummyPin<L> {
+        type Error = Infallible;
+
+        fn try_set_high(&mut self) -> Result<(), Self::Error> {
+            Ok(())
+        }
+
+        fn try_set_low(&mut self) -> Result<(), Self::Error> {
+            Ok(())
+        }
+    }
+
+    impl StatefulOutputPin for DummyPin<level::Low> {
+        fn try_is_set_high(&self) -> Result<bool, Self::Error> {
+            Ok(false)
+        }
+
+        fn try_is_set_low(&self) -> Result<bool, Self::Error> {
+            Ok(true)
+        }
+    }
+
+    impl StatefulOutputPin for DummyPin<level::High> {
+        fn try_is_set_high(&self) -> Result<bool, Self::Error> {
+            Ok(true)
+        }
+
+        fn try_is_set_low(&self) -> Result<bool, Self::Error> {
+            Ok(false)
+        }
+    }
```

Comment:
Shouldn't a stateful pin really be stateful?

Reply:
The pin could be made to remember what the last state set but then a boolean needs to be added to the struct and the methods are not zero-cost anymore.
If we want to keep the output pin implementation zero-cost, we could also create a separate `DummyStatefulOutputPin` which adds the boolean.

Reply:
I don't think it makes sense to implement `StatefulOutputPin` on something which is not stateful as that's really the only purpose of that trait.

Reply:
Ok, I could also just remove the `StatefulOutputPin` implementation. The original motivation issue did not include it anyway.

Reply:
And conversely `toggleable` doesn't make sense on a stateless pin either.

Reply:
alright I removed them.

Reply:
Works for me, alternatively we could do as you suggested and add another non-zero-cost stateful pin although I'm not super convinced whether that is really useful.

Reply:
Yeah I think we can go with just this and add a stateful pin later on if somebody asks for it.